### PR TITLE
Refactor Makefile.toml and use it as the source of truth for (almost) all CI

### LINKED ISSUE #612: Refactor Makefile.toml into smaller files
Makefile.toml is currently a grab-bag of a lot of different build tasks, and it keeps growing.  It would be better if we could refactor separate parts of it into separate files that get included into the main file.

I believe that this would also enable us to use Duckscript functions, allowing us to reduce code duplication.

This could involve a new top-level directory named `build_scripts` or similar.

### PR DESCRIPTION
**Pull request by Manishearth**
This refactors Makefile.toml into multiple subfiles. It also adds various `ci-job-foo` make tasks which are directly run by CI, so adding further CI can be done by tweaking Makefile.toml only.

This reduces the number of parallel CI jobs. We still have separate jobs for things that are fast-fail: there's a separate check, tidy, fmt, and clippy job, but otherwise there's a "test" (for most tests), "features" (for the "all feature combos" test), and "wasm" jobs. Jobs like "bincode" have been rolled into "test": unless the job has a reason to be long running or require weird dependencies, it should be a part of "test".

Fixes #612

### CODE REVIEW

File: .cargo/config.toml
Comment (sffc) [inline .cargo/config.toml:10]: Nit: Instead of updating these aliases to invoke `cargo make`, please just delete them.  They were here previously to allow Makefile.toml and CI to execute the same command.  But now if CI invokes `cargo make` directly, we don't need these, and they will only cause confusion.

File: .github/workflows/build-test.yml
Comment (sffc) [inline .github/workflows/build-test.yml:31]: Issue: You probably need to install `cargo make` in any CI in which you want to use it.

File: .cargo/config.toml
Comment (Manishearth) [reply .cargo/config.toml:10]: Great, I was going to ask about this anyway :smile: 

I didn't want to break any workflows

**Comment by Manishearth** [comment]
image

looks better now!

File: .cargo/config.toml
Comment (sffc) [inline .cargo/config.toml:39]: If you delete these two metatasks, please update the corresponding docs in CONTRIBUTING.md.

File: CONTRIBUTING.md
Comment (sffc) [inline CONTRIBUTING.md:37]: Question: What's the use case for a developer running `cargo tidy`?

**Review by sffc — APPROVED** [review]

File: CONTRIBUTING.md
Comment (Manishearth) [reply CONTRIBUTING.md:37]: because usually devs will have run the other checks (build/fmt) but often forget to run the license check/etc

I'm also okay with a tidy task that is fmt+license+readme

Comment (sffc) [reply CONTRIBUTING.md:37]: I think adding fmt would be most useful here.  I feel like the most common error is fmt, followed by license.

**Comment by Manishearth** [comment]
I made it cache cargo-make, but I plan to write a github action that makes that easy to do. The existing install GHA supports caching, but badly (it uses an external cache), and using the builtin cache is blocked on https://github.com/actions-rs/install/pull/5

File: CONTRIBUTING.md
Comment (Manishearth) [reply CONTRIBUTING.md]: Done

**Comment by Manishearth** [comment]
This is ready to land, though before we land this we need to update the required CI runs in the repo settings

**Comment by Manishearth** [comment]
I tried cleaning up the cargo make installs in https://github.com/unicode-org/icu4x/pull/787, but it didn't work. Eventually actions-rs/install will get support for the native cache.

File: tools/scripts/tidy.toml
Comment (sffc) [inline tools/scripts/tidy.toml:7]: Suggestion (optional): Delete the "tidy-minus-fmt" task, move these up into "ci-job-tidy", and make "tidy" invoke fmt + ci-job-tidy.

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
When we're ready to merge this, we should change the branch permissions to reflect the new CI.

File: tools/scripts/tidy.toml
Comment (Manishearth) [reply tools/scripts/tidy.toml:7]: I kinda want the ci-job-foo to be toplevel so you can change CI jobs without changing other stuff.

Comment (Manishearth) [reply tools/scripts/tidy.toml]: eh, fixed it anyway

**Review by dminor — APPROVED** [review]

**Comment by Manishearth** [comment]
Seems like I have to change the rules _before_ the push, so we have to wait for these checks to pass and I need another r+

**Pull request merged**
